src/train.jl

```diff
@@ -0,0 +1,136 @@
+    end
+    state
+end
+
+"""
+    train!(loss, model, data, opt)
+
+Uses a `loss` function and training `data` to improve the `model`'s parameters
+according to a particular optimisation rule `opt`. Iterates through `data` once,
+evaluating `loss(model, d...)` for each `d` in data.
+
+For example, with these definitions...
+```
+data = [(x1, y1), (x2, y2), (x3, y3)];  # each element must be a tuple
+
+loss3(m, x, y) = norm(m(x) .- y)        # the model is the first argument
+
+opt = Flux.setup(Adam(), model)         # explicit setup of optimiser momenta
+```
+...calling `Flux.train!(loss3, model, data, opt)` runs a loop much like this,
+using Zygote's "explicit" mode for the gradient:
+```
+for d in data
+    ∂L∂m = gradient(loss3, model, d...)[1]
+    update!(opt, model, ∂L∂m)           # method for "explicit" gradient
+end
+```
+You can also write this loop yourself, if you need more flexibility.
+For this reason `train!` is not highly extensible.
+It adds only a few featurs to the loop above:
+
+* Stop with a `DomainError` if the loss is infinite or `NaN` at any point.
+
+* Show a progress bar using [`@withprogress`](https://github.com/JuliaLogging/ProgressLogging.jl).
+
+Note that the built-in loss functions accept 3 arguments, allowing for instance
+`train!(Flux.Losses.mse, model, data, opt)` instead of defining `loss3` as above.
+
+!!! note
+    This method has significant changes from the one in Flux ≤ 0.13:
+    * It now takes the `model` itself, not the result of [`Flux.params`](@ref).
+      (This is to move away from Zygote's "implicit" parameter handling, with `Grads`.)
+    * Instead of `loss` being a function which typically accepts two arguments
+      (the input `x` and expected output `y` from each element of `data`)
+      now it should typically accept three, the first of which is the `model` itself.
+    * `data` must iterate tuples, otherwise you get an error.
+      (Previously non-tuple types were not splatted into the loss. 
+      Pass in `((d,) for d in data)` to simulate this.)
+    * `opt` should be the result of [`Flux.setup`](@ref). Using an optimiser
+      such as `Adam()` without this step should give you a warning.
+    * Callback functions are not supported.
+      But any code can be included in the above `for` loop.
+"""
+function train!(loss, model, data, opt; cb = nothing)
+  isnothing(cb) || error("""train! does not support callback functions.
+                            For more control use a loop with `gradient` and `update!`.""")
+  losses = Float32[]
+  @withprogress for (i,d) in enumerate(data)
+    d isa Tuple || error("""train! expects as data an iterator producing tuples, but got $(typeof(d)).
+                            Pass it `((d,) for d in data)`, or use `gradient` and `update!` for more control.""")
+    # l, (g, _...) = explicit_withgradient(loss, model, d...)  # BTW this un-thunks gradient w.r.t. data. Could avoid that
+    l, (g, _...) = explicit_withgradient(m -> loss(m, d...), model)
```

Comment:
If we consider this function to be scrutinized by many people which will then move on to define their own training loop, maybe better to be really explicit and do
```suggestion
    l, (g, _...) = Zygote.withgradient(m -> loss(m, d...), model)
```

Reply:
Should probably go.

Earlier commits had a mechanism like `@train_autodiff Tracker` which replaced this function, as a global switch of what AD to use. (Turing does something like this.) But... maybe this too I think is better done by writing a loop and calling `Tracker.withgradient` or whatever.